test/unit/local/steps/incomplete_fixer.js

```diff
@@ -0,0 +1,116 @@
+
+const completionChanges = async (buffer) => buffer.pop().map(completedEvent)
+
+describe('local/steps/initialDiff', () => {
+  let builders
+  let syncDir
+  let buffer
+  let options
+  let initialScanDone
+
+  const CHECKSUM = 'checksum'
+  const checksumer = {
+    push: async path => CHECKSUM,
+    kill: () => {}
+  }
+
+  before('instanciate config', configHelpers.createConfig)
+  before(function () {
+    options = { syncPath: this.config.syncPath, checksumer }
+    syncDir = new ContextDir(this.config.syncPath)
+  })
+  beforeEach('instanciate pouch', pouchHelpers.createDatabase)
+  beforeEach('populate pouch with documents', function () {
+    builders = new Builders({pouch: this.pouch})
+    buffer = new Buffer()
+    initialScanDone = builders.event().action('initial-scan-done').path('').build()
+  })
+  afterEach('clean pouch', pouchHelpers.cleanDatabase)
+  after('clean config directory', configHelpers.cleanConfig)
+
+  context('without any complete "renamed" event', () => {
+    it('drops incomplete events', async function () {
+      const events = [
+        builders.event().incomplete().action('created').path('foo1').build(),
+        builders.event().incomplete().action('modified').path('foo2').build(),
+        builders.event().incomplete().action('deleted').path('foo3').build(),
+        builders.event().incomplete().action('scan').path('foo4').build()
+      ]
+      buffer.push(_.cloneDeep(events))
+      buffer.push(_.cloneDeep([initialScanDone]))
+
+      const completedBuffer = await incompleteFixer.loop(buffer, options)
+      should(await completedBuffer.pop()).deepEqual([initialScanDone])
+    })
+  })
+
+  context('with a complete "renamed" event in the batch', () => {
+    it('leaves complete events untouched', async function () {
+      const events = [
+        builders.event().action('created').path('file').build(),
+        builders.event().action('renamed').oldPath('file').path('foo').build()
+      ]
+      buffer.push(_.cloneDeep(events))
+
+      const completedBuffer = await incompleteFixer.loop(buffer, options)
+      should(await completedBuffer.pop()).deepEqual(events)
+    })
+
+    it('rebuilds the first incomplete event matching the "renamed" event old path', async function () {
+      const events = [
+        builders.event().incomplete().kind('file').action('created').path('src/foo1').build(),
+        builders.event().incomplete().kind('file').action('modified').path('src/foo2').build(),
+        builders.event().incomplete().kind('file').action('deleted').path('src/foo3').build(),
+        builders.event().incomplete().kind('file').action('renamed').oldPath('src/foo4').path('foo').build(),
+        builders.event().incomplete().kind('file').action('scan').path('src/foo5').build()
+      ]
+      const renamed = builders.event().kind('directory').action('renamed').oldPath('src').path('dst').build()
+      buffer.push(_.cloneDeep(events))
+      buffer.push(_.cloneDeep([renamed, initialScanDone]))
```

Comment:
Test description says `with a complete "renamed" event in the batch`.
But the event is actually in a separate batch... Does it matter?

Reply:
It was in the same batch but I moved it to another batch for convenience when I introduced the named events.